### ISSUE TITLE
don't actually build .a files on Linux, instead pass the original .o files to the final link

### ISSUE THREAD

**Issue by bsilver8192** [bazelbuild/bazel#1411]
In #492, @ulfjack mentioned that "We don't actually build .a files on Linux, where we instead pass the original .o files to the final link (or at least, we have code to do that, and I think it's enabled by default). This is faster than building intermediate .a files and uses less disk space." That sounds nice, but doesn't seem to actually work:

``` console
$ bazel build --subcommands //:main
INFO: Found 1 target...
>>>>> # //:main [action 'Creating runfiles tree bazel-out/local-fastbuild/bin/main.runfiles']
...
>>>>> # //:gen_files [action 'Executing genrule //:gen_files']
...
>>>>> # //:a [action 'Compiling a.c']
...
>>>>> # //:main [action 'Compiling main.c']
...
>>>>> # //:a [action 'Linking liba.a']
...
>>>>> # //:main [action 'Linking main']
...
Target //:main up-to-date:
  bazel-bin/main
INFO: Elapsed time: 2.655s, Critical Path: 0.61s
```

BUILD:

``` python
genrule(
  name = 'gen_files',
  outs = ['a.c', 'main.c'],
  cmd = 'echo "int main() {}" > $(location main.c) && touch $(location a.c)',
)

cc_library(
  name = 'a',
  srcs = ['a.c'],
)

cc_binary(
  name = 'main',
  srcs = ['main.c'],
  deps = [':a'],
)
```

**Comment by ulfjack** [comment]
The CROSSTOOL file needs to declare that it supports that mode by setting
supports_start_end_lib = true. Maybe the cc_configure rule doesn't do that
by default? Maybe it depends on the compiler version?

**Comment by bsilver8192** [comment]
That would be the problem. Those appear to be a gold-only flags, and cc_configure.bzl doesn't attempt to auto-detect using gold. If I add `linker_flag: "-fuse-ld=gold"` (necessary on my Debian Jessie system to use gold) and change to `supports_start_end_lib: true`, then it works.

Would it make sense to have cc_configure.bzl try -fuse-ld=gold and set stuff appropriately if it works? Or maybe somehow detect if the default linker is gold?

**Comment by philwo** [comment]
Topic area is C++ auto configuration, so assignee is Damien. :)

**Comment by dslomov** [comment]
@damienmg Any more information that is needed here?

**Comment by damienmg** [comment]
Sorry lost track with my personnal issues, it is indeed something that we should do, triaged.

**Comment by mhlopko** [comment]
Should be fixed by cc3d79cca775cad9e9193d38bedf4df51a5c8a50, closing. Thank you bsilver8192!
